clippy_lints/src/loops.rs

```diff
@@ -2112,3 +2139,101 @@ fn path_name(e: &Expr) -> Option<Name> {
     };
     None
 }
+
+fn check_infinite_loop<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, cond: &'tcx Expr, block: &'tcx Block, expr: &'tcx Expr) {
+    let mut mut_var_visitor = MutableVarsVisitor {
+        cx,
+        ids: HashSet::new(),
+        skip: false,
+    };
+    walk_expr(&mut mut_var_visitor, expr);
+    if mut_var_visitor.skip {
+        return;
+    }
+
+    if mut_var_visitor.ids.len() == 0 {
+        span_lint(
+            cx,
+            WHILE_IMMUTABLE_CONDITION,
+            cond.span,
+            "all variables in condition are immutable. This might lead to infinite loops.",
+        );
+        return;
+    }
+
+    let mut use_visitor = MutablyUsedVisitor {
+        cx,
+        ids: mut_var_visitor.ids,
+        any_used: false,
+    };
+    walk_block(&mut use_visitor, block);
+    if !use_visitor.any_used {
+        span_lint(
+            cx,
+            WHILE_IMMUTABLE_CONDITION,
+            expr.span,
+            "Variable in the condition are not mutated in the loop body. This might lead to infinite loops.",
+        );
+    }
+}
+
+/// Collects the set of mutable variable in an expression
+/// Stops analysis if a function call is found
+struct MutableVarsVisitor<'a, 'tcx: 'a> {
+    cx: &'a LateContext<'a, 'tcx>,
+    ids: HashSet<NodeId>,
+    skip: bool,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for MutableVarsVisitor<'a, 'tcx> {
+    fn visit_expr(&mut self, ex: &'tcx Expr) {
+        match ex.node {
+            ExprPath(_) => if let Some(node_id) = check_for_mutability(self.cx, &ex) {
+                self.ids.insert(node_id);
+            },
+
+            // If there is any fuction/method call… we just stop analysis
+            ExprCall(..) | ExprMethodCall(..) => self.skip = true,
+
+            _ => walk_expr(self, ex),
+        }
+    }
+
+    fn visit_block(&mut self, _b: &'tcx Block) {}
+
+    fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'tcx> {
+        NestedVisitorMap::None
+    }
+}
+
+/// checks within an expression/statement if any of the variables are used mutably
+struct MutablyUsedVisitor<'a, 'tcx: 'a> {
+    cx: &'a LateContext<'a, 'tcx>,
+    ids: HashSet<NodeId>,
+    any_used: bool,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for MutablyUsedVisitor<'a, 'tcx> {
```

Comment:
I just noticed there is [ExprUseVisitor](https://michael-f-bryan.github.io/rustc-internal-docs/rustc/middle/expr_use_visitor/struct.ExprUseVisitor.html).
Would it be safer to use or is the match on AddrOf and Assign expressions enough?
I am actually pretty confused by all the arguments required by ExprUseVisitor as well as what Delegate is suppose to do in consume/borrow methods…